src/legacy/ui/public/index_patterns/_field.ts

```diff
@@ -0,0 +1,192 @@
+    // extends the Field.prototype but gets it's properties
+    // defined using the logic below
+    const obj = new ObjDefine(spec, Field.prototype);
+
+    if (spec.name === '_source') {
+      spec.type = '_source';
+    }
+
+    // find the type for this field, fallback to unknown type
+    let type = getKbnFieldType(spec.type);
+    if (spec.type && !type) {
+      const title = i18n.translate('common.ui.indexPattern.unknownFieldHeader', {
+        values: { type: spec.type },
+        defaultMessage: 'Unknown field type {type}',
+      });
+      const text = i18n.translate('common.ui.indexPattern.unknownFieldErrorMessage', {
+        values: { name: spec.name, title: indexPattern.title },
+        defaultMessage: 'Field {name} in indexPattern {title} is using an unknown field type.',
+      });
+
+      toastNotifications.addDanger({
+        title,
+        text,
+      });
+    }
+
+    if (!type) type = getKbnFieldType('unknown');
+
+    let format = spec.format;
+    if (!format || !(format instanceof FieldFormat)) {
+      format =
+        indexPattern.fieldFormatMap[spec.name] ||
+        fieldFormats.getDefaultInstance(spec.type, spec.esTypes);
+    }
+
+    const indexed = !!spec.indexed;
+    const scripted = !!spec.scripted;
+    const searchable = !!spec.searchable || scripted;
+    const aggregatable = !!spec.aggregatable || scripted;
+    const readFromDocValues = !!spec.readFromDocValues && !scripted;
+    const sortable = spec.name === '_score' || ((indexed || aggregatable) && type.sortable);
+    const filterable =
+      spec.name === '_id' || scripted || ((indexed || searchable) && type.filterable);
+    const visualizable = aggregatable;
+
+    this.name = '';
+    obj.fact('name');
+    this.type = '';
+    obj.fact('type');
+    obj.fact('esTypes');
+    obj.writ('count', spec.count || 0);
+
+    // scripted objs
+    obj.fact('scripted', scripted);
+    obj.writ('script', scripted ? spec.script : null);
+    obj.writ('lang', scripted ? spec.lang || 'painless' : null);
+
+    // stats
+    obj.fact('searchable', searchable);
+    obj.fact('aggregatable', aggregatable);
+    obj.fact('readFromDocValues', readFromDocValues);
+
+    // usage flags, read-only and won't be saved
+    obj.comp('format', format);
+    obj.comp('sortable', sortable);
+    obj.comp('filterable', filterable);
+    obj.comp('visualizable', visualizable);
+
+    // computed values
+    obj.comp('indexPattern', indexPattern);
+    obj.comp('displayName', shortDotsEnable ? shortenDottedString(spec.name) : spec.name);
+    this.$$spec = spec;
+    obj.comp('$$spec', spec);
+
+    // conflict info
+    obj.writ('conflictDescriptions');
+
+    // multi info
+    obj.fact('parent');
+    obj.fact('subType');
+
+    return obj.create();
+  }
+
+  /** @deprecated */
+  public get indexed() {
+    throw new Error(
+      'field.indexed has been removed, see https://github.com/elastic/kibana/pull/11969'
```

Comment:
yeah i think we should be able to remove those by now, but would prefer to do it separately to avoid mixing more api changes with a TS-focused PR. i've made a note to look at that next